client/utils/codemirror-search.js

```diff
@@ -0,0 +1,424 @@
+
+
+    } else {
+      searchField.focus();
+    }
+  }
+
+  function dialog(cm, text, shortText, deflt, f) {
+    if (cm.openDialog) cm.openDialog(text, f, {value: deflt, selectValueOnOpen: true});
+    else f(prompt(shortText, deflt));
+  }
+
+  var lastSelectedIndex = 0;
+  function confirmDialog(cm, text, shortText, fs) {
+    if (cm.openConfirm) cm.openConfirm(text, fs);
+    else if (confirm(shortText)) fs[0]();
+
+    var dialog = document.getElementsByClassName("CodeMirror-dialog")[0];
+    var buttons = dialog.getElementsByTagName("button");
+    buttons[lastSelectedIndex].focus();
+    for (var i = 0; i < buttons.length; i += 1) {
+      (function (index) {
+        var button = buttons[index];
+        button.addEventListener("focus", function (e) {
+          lastSelectedIndex = index === buttons.length - 1 ? 0 : index;
+        });
+        button.addEventListener("keyup", function (e) {
+          if (e.keyCode === 37) { // arrow left
+            var prevButton = index === 0 ? buttons.length - 1 : index - 1;
+            buttons[prevButton].focus();
+          }
+          if (e.keyCode === 39) { // arrow right
+            var nextButton = index === buttons.length - 1 ? 0 : index + 1;
+            buttons[nextButton].focus();
+          }
+          if (e.keyCode === 27) { // esc
+            cm.focus();
+          }
+        });
+        button.addEventListener("click", function () {
+          if (index === buttons.length - 1) { // "done"
+            lastSelectedIndex = 0;
+          }
+        })
+      })(i);
+    }
+  }
+
+  function parseString(string) {
+    return string.replace(/\\(.)/g, function(_, ch) {
+      if (ch == "n") return "\n"
+      if (ch == "r") return "\r"
+      return ch
+    })
+  }
+
+  /*
+    Parses the query text and state and returns
+    a RegExp ready for searching
+  */
+  function parseQuery(query, state) {
+    var emptyQuery = 'x^'; // matches nothing
+
+    if (query === '') { // empty string matches nothing
+      query = emptyQuery;
+    } else {
+      if (state.regexp === false) {
+        query = parseString(query);
+        query = query.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+      }
+
+      if (state.wholeWord) {
+        query += '\\b';
+      }
+    }
+
+    var regexp;
+
+    try {
+      regexp = new RegExp(query, state.caseInsensitive ? "gi" : "g");
+    } catch (e) {
+      regexp = new RegExp(emptyQuery, 'g');
+    }
+
+    // If the resulting regexp will match everything, do not use it
+    if (regexp.test('')) {
+      return new RegExp(emptyQuery, 'g');
+    }
+
+    return regexp;
+  }
+
+  var queryDialog = `
+    <h3 class="CodeMirror-search-title">Find</h3>
+    <input type="text" class="search-input CodeMirror-search-field" placeholder="Search for" />
+    <div class="CodeMirror-search-actions">
+      <div class="CodeMirror-search-modifiers button-wrap">
+        <button
```

Comment:
### Selecting options

These can all be used together. 

- Case Sensitive
- Whole word
- Regex: Does a regex search
- Regex + Case Sensitive: Adds `/i` flag
- Regex + Whole word: Adds `\b` to expression

### Checkbox instead of buttons

Makes sense to me. Are there any examples in the code of using a styled checkbox input? 

Alternatively, could I use the [`aria-checked` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_checkbox_role)?